Initial Commit

IIExpanse opened the pull request:
Андрей, добрый день!
Застрял дольше обычного, пока копался во всем новом функционале и библиотеках.

В тестах осталась одна странная ошибка, связанная с непонятным ответом сервера (SocketException: Unexpected end of file from server). Провоцируется только, когда проблемный тест shouldNotUpdateNonexistentFilm из FilmControllerTest исполняется вместе с другими тестами, использующими метод exchange класса TestRestTemplate. 

Сам по себе тест проходит успешно. Изменение настроек сервера и другие манипуляции пока не помогли. Что интересно, тест из UserControllerTest с аналогичным функционалом для контроллера пользователей проходит без проблем.
Я ещё попытаюсь устранить ошибку, но поскольку уже время поджимает, пока отсылаю так.

IIExpanse committed:
Initial Commit

shargin commented on pom.xml:
```diff
@@ -0,0 +1,53 @@
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-test</artifactId>
+			<scope>test</scope>
```
Зависимости из скоупа тест лучше размещать в конце

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,70 @@
+
+    @PostMapping
+    public ResponseEntity<Film> addNewFilm(@Valid @RequestBody Film film) throws ValidationException {
+        int id = idCounter++;
```
Генерацию id лучше выполнять в отдельном методе

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,70 @@
+    public ResponseEntity<Film> addNewFilm(@Valid @RequestBody Film film) throws ValidationException {
+        int id = idCounter++;
+        film.setId(id);
+        boolean isPresent = filmsMap.containsKey(id);
```
Лучше не позволять пользователям иметь возможность манипулировать id, лучше игнорировать их

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,70 @@
+            return new ResponseEntity<>(filmsMap.get(id), HttpStatus.CREATED);
+
+        } else {
+            String errorMessage = String.format("Добавляемый фильм с id=%d уже существует.", id);
```
slf4j поддерживает плейсхолдеры {}

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,70 @@
+        }
+    }
+
+    @PutMapping
```
Чаще всего в методе для изменения id передают в url, например, PUT /users/{id}

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,70 @@
+        if (filmsMap.containsKey(id)) {
+            filmsMap.put(id, film);
+            log.debug("Обновлена информация о фильме: {}", film);
+            return new ResponseEntity<>(filmsMap.get(id), HttpStatus.OK);
```
Тут, видимо, можно обойтись без get, т.к. возвращается тот же объект, который был прислан

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,70 @@
+
+    @GetMapping
+    public ResponseEntity<List<Film>> getFilmsList() {
+        return new ResponseEntity<>(List.copyOf(filmsMap.values()), HttpStatus.OK);
```
ResponseEntity.ok(объект)

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,79 @@
+@RestController
+@RequestMapping("/users")
+@Slf4j
+public class UserController {
```
Тут аналогично предыдущему контроллеру

shargin commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,49 @@
+    @Positive
+    private final int duration;
+
+    @JsonCreator
```
Думаю, тут не требуется creator, т.к. все параметры из json соответствуют полям класса. Т.к. подключен ломбок, то можно воспользоваться аннотациями  AllArgsConstructor, NoArgsConstructor

shargin commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,49 @@
+public class User {
+
+    private int id;
+    @Email
```
Email не запрещает пустые строки

shargin commented on src/main/java/ru/yandex/practicum/filmorate/validators/impl/NoWhitespaceValidator.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        return !value.contains(" ");
```
Нет проверки на null у value

shargin commented on src/main/java/ru/yandex/practicum/filmorate/validators/impl/ReleaseDateValidator.java:
```diff
@@ -0,0 +1,22 @@
+
+public class ReleaseDateValidator implements ConstraintValidator<ReleaseDateConstraint, LocalDate> {
+
+    public static final LocalDate CINEMA_BIRTH_DAY = LocalDate.of(1895, 12, 28);
```
Раз есть желание использовать валидаторы, то имеет смысл сделать его более универсальным, т.е. передавать дату в аннотацию, а не использовать какую-то конкретную

shargin commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -0,0 +1,15 @@
+@SpringBootTest
+public class FilmorateApplicationTests {
+
+	public static final String BASE_URL = "http://localhost:8080";
```
Это лучше перенести в соответствующий тест-класс

shargin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,175 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
```
Лучше использовать RANDOM_PORT, так тесты не будут зависеть запущен ли еще экземпляр на данном порту, а порт на котором запустилось приложение можно узнать, например
@LocalServerPort
 int port;
И решится проблема с тестом

shargin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,175 @@
+    private static final String context = "/films";
+    private static final String testURI = FilmorateApplicationTests.BASE_URL + context;
+
+    private final TestRestTemplate restTemplate = new TestRestTemplate();
```
Есть соответствующий бин, создавать ничего не требуется

IIExpanse committed:
fix: 1. Тестовые зависимости в Pom перемещены вниз.
2. Из взаимодействий с id объектов, получаемых для добавления, осталось только присваивание генерируемых сервером.
3. В добавлении логов используются только плейсхолдеры.
4. Id передается в url.
5. POST запросам возвращается присланный объект.
6. Для GET запросов используется ResponseEntity.ok(объект).
7. Из Film и User убран JsonCreator.
8. Email проверяется на пустую строку.
9. В NoWhitespaceConstraint добавлена проверка значения на null.
10. в ReleaseDateConstraint дата передается через аннотацию.
11. Переделаны поля тестовых классов: port задается через RANDOM_PORT, TestRestTemplate через бин.

IIExpanse committed:
Fix: Корректировки URL PUT запросов для прохождения встроенных тестов (убран слэш).

Nagash40 commented:
Отправляю на следующее ревью. Я изменил передачу id на передачу через url вместо объекта согласно комментарию, но местные тесты постмана похоже передают id через объект.

shargin commented on src/main/java/ru/yandex/practicum/filmorate/validators/ReleaseDateConstraint.java:
```diff
@@ -0,0 +1,24 @@
+
+    Class<? extends Payload>[] payload() default {};
+
+    String[] earliestPossibleDate() default {};
```
Со строкой-датой получается не очень надежно, т.к. неясно в каком формате передавать, с каким разделителями с минутами, часами и т.д. Лучше отдельно завести параметры для для, месяца и года. Так шансов сделать ошибку бедет меньше

shargin commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @PutMapping
+    public ResponseEntity<User> updateUser(@Valid @RequestBody User user, @RequestParam int id) throws ValidationException {
```
Да, действительно постман-тесты ожидают ид в объекте, надо исправить, чтобы прошли тесты

IIExpanse committed:
Fix: 1. В аннотацию ReleaseDateConstraint теперь передаются отдельные целочисленные значения года, месяца и дня.
2. Возвращена передача id в объекте.

Nagash40 commented:
Изменил передачу стринги в аннотацию на передачу целочисленных значений года, месяца и дня по отдельности и вернул передачу id через объект.
Тесты наконец прошли.
